Remove USE_CUDA from ProcessGroupGloo

lw opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59563 Move c10d to libtorch(_cuda)
* #59697 Add guards for USE_C10D_FOO in relevant c10d files
* #59696 Fix Windows guards in c10d
* #59562 Remove USE_CUDA from c10d reducer/logger
* **#59561 Remove USE_CUDA from ProcessGroupGloo**

Differential Revision: [D28931379](https://our.internmc.facebook.com/intern/diff/D28931379/)

lw committed:
Remove USE_CUDA from ProcessGroupGloo



Differential Revision: [D28931379](https://our.internmc.facebook.com/intern/diff/D28931379/)

[ghstack-poisoned]

lw committed:
Update on "Remove USE_CUDA from ProcessGroupGloo"


Differential Revision: [D28931379](https://our.internmc.facebook.com/intern/diff/D28931379/)

[ghstack-poisoned]

lw committed:
Update on "Remove USE_CUDA from ProcessGroupGloo"


Differential Revision: [D28931379](https://our.internmc.facebook.com/intern/diff/D28931379/)

[ghstack-poisoned]

lw committed:
Update on "Remove USE_CUDA from ProcessGroupGloo"


Differential Revision: [D28931379](https://our.internmc.facebook.com/intern/diff/D28931379/)

[ghstack-poisoned]

lw committed:
Update on "Remove USE_CUDA from ProcessGroupGloo"


Differential Revision: [D28931379](https://our.internmc.facebook.com/intern/diff/D28931379/)

[ghstack-poisoned]

lw committed:
Update on "Remove USE_CUDA from ProcessGroupGloo"


Differential Revision: [D28931379](https://our.internmc.facebook.com/intern/diff/D28931379/)

[ghstack-poisoned]

lw committed:
Update on "Remove USE_CUDA from ProcessGroupGloo"


Differential Revision: [D28931379](https://our.internmc.facebook.com/intern/diff/D28931379/)

[ghstack-poisoned]

lw committed:
Update on "Remove USE_CUDA from ProcessGroupGloo"


Differential Revision: [D28931379](https://our.internmc.facebook.com/intern/diff/D28931379/)

[ghstack-poisoned]

lw committed:
Update on "Remove USE_CUDA from ProcessGroupGloo"


Differential Revision: [D28931379](https://our.internmc.facebook.com/intern/diff/D28931379/)

[ghstack-poisoned]

lw committed:
Update on "Remove USE_CUDA from ProcessGroupGloo"


Differential Revision: [D28931379](https://our.internmc.facebook.com/intern/diff/D28931379/)

[ghstack-poisoned]

lw committed:
Update on "Remove USE_CUDA from ProcessGroupGloo"


Differential Revision: [D28931379](https://our.internmc.facebook.com/intern/diff/D28931379/)

[ghstack-poisoned]

lw committed:
Update on "Remove USE_CUDA from ProcessGroupGloo"


Differential Revision: [D28931379](https://our.internmc.facebook.com/intern/diff/D28931379/)

[ghstack-poisoned]

lw committed:
Update on "Remove USE_CUDA from ProcessGroupGloo"


Differential Revision: [D28931379](https://our.internmc.facebook.com/intern/diff/D28931379/)

[ghstack-poisoned]

lw committed:
Update on "Remove USE_CUDA from ProcessGroupGloo"


Differential Revision: [D28931379](https://our.internmc.facebook.com/intern/diff/D28931379/)

[ghstack-poisoned]

lw commented on c10/cuda/CUDAStream.h:
```diff
@@ -67,7 +68,12 @@ class C10_CUDA_API CUDAStream {
   /// and will raise an error if the Stream is not, in fact, a CUDA stream.
   explicit CUDAStream(Stream stream) : stream_(stream) {
-    TORCH_CHECK(stream_.device_type() == DeviceType::CUDA);
+    TORCH_CHECK(
+        stream_.device_type() == DeviceType::CUDA,
+        "Got device type ",
+        stream_.device_type(),
+        "\n",
+        c10::get_backtrace());
```
This was a change to help me debug some ROCm failures, I'll undo it before landing

lw committed:
Update on "Remove USE_CUDA from ProcessGroupGloo"


Differential Revision: [D28931379](https://our.internmc.facebook.com/intern/diff/D28931379/)

[ghstack-poisoned]

lw committed:
Update on "Remove USE_CUDA from ProcessGroupGloo"


Differential Revision: [D28931379](https://our.internmc.facebook.com/intern/diff/D28931379/)

[ghstack-poisoned]

agolynski commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -920,9 +895,9 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupGloo::broadcast(
+      break;
     case at::kCUDA:
-#endif
+      TORCH_INTERNAL_ASSERT(c10::impl::hasDeviceGuardImpl(at::kCUDA));
```
1) nit: use .hasCUDA()?
2) let's preserve the old behavior with invalidArgument instead of checkfail?

lw replied:
> let's preserve the old behavior with invalidArgument instead of checkfail?

The reason I put an assert here is because this should never ever possibly occur: if the user has been able to give us a CUDA tensor, then CUDA _must_ be loaded and that check _must_ pass. In fact, I wouldn't even oppose removing that assert entirely, since in a sense it's pointless...

I'll replace this with `at::hasCUDA()`, which is more concise and I think will give the same answer, although technically it does behave somewhat differently: PyTorch apparently has two "registries" for CUDA to plug in, one is the CUDAHooks (which is what is checked by `at::hasCUDA()`) the other one is the DeviceGuardImpl registry (which is used by c10::Events and such, which is what we actually care about here).

agolynski commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -1421,9 +1384,9 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupGloo::allreduce(
+      break;
     case at::kCUDA:
-#endif
+      TORCH_INTERNAL_ASSERT(c10::impl::hasDeviceGuardImpl(at::kCUDA));
```
same here

agolynski commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -1871,9 +1815,9 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupGloo::allgather(
+      break;
     case at::kCUDA:
-#endif
+      TORCH_INTERNAL_ASSERT(c10::impl::hasDeviceGuardImpl(at::kCUDA));
```
also here

agolynski commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -2123,24 +2063,21 @@ class AsyncGatherCUDAWork : public AsyncGatherWork {
 
   void run() override {
     // Synchronize with copy operations.
-    at::cuda::OptionalCUDAGuard device_guard;
```
no need to switch devices here? was this only necessary for cuda API?

lw replied:
I don't see any operation below that seems to depend on the current device, do you?

agolynski commented on torch/lib/c10d/ProcessGroupGloo.cpp:
```diff
@@ -2208,9 +2142,9 @@ c10::intrusive_ptr<ProcessGroup::Work> ProcessGroupGloo::gather(
+      break;
     case at::kCUDA:
-#endif
+      TORCH_INTERNAL_ASSERT(c10::impl::hasDeviceGuardImpl(at::kCUDA));
```
and here

agolynski reviewed (approved):
Nice work, this PG GLOO looks much cleaner now!

LG with some comments

lw committed:
Update on "Remove USE_CUDA from ProcessGroupGloo"


Differential Revision: [D28931379](https://our.internmc.facebook.com/intern/diff/D28931379/)

[ghstack-poisoned]
